corgi/core/constants.py

```diff
@@ -69,3 +72,117 @@
         "Red Hat build of Quarkus Middleware",
     ],
 }
+
+# Strings inside tuples are automatically joined together onto one line
+# So the SQL syntax will be correct in the final stored procedure code
+LATEST_FILTER_DEFINITION = (
+    "get_latest_component( "
+    "model_type text, ps_ofuri text, component_type text, component_ns text, "
+    "component_name text, component_arch text, include_inactive_streams boolean) "
+    "RETURNS uuid AS $$"
+)
+LATEST_FILTER_FIELDS = (
+    "core_component.uuid, core_component.epoch, "
+    "core_component.version, core_component.release from core_component"
+)
+LATEST_FILTER_WHERE = (
+    "WHERE core_component.name=component_name "
+    "AND core_component.namespace=component_ns "
+    "AND core_component.type=component_type "
+    "AND core_component.arch=component_arch "
+    "AND (include_inactive_streams OR core_productstream.active) "
+    f"AND ({ROOT_COMPONENTS_SQL})"
+)
+
+LATEST_FILTER_INTO = (
+    "INTO component_uuid, component_epoch, component_version, component_release; "
+    "EXIT WHEN NOT FOUND; "
+    "IF rpmvercmp_epoch(component_epoch, component_version, component_release, "
+    "latest_epoch, latest_version, latest_release) >= 0 THEN "
+    "latest_uuid := component_uuid; "
+    "latest_epoch := component_epoch; "
+    "latest_version := component_version; "
+    "latest_release := component_release; "
+    "END IF;"
+)
+
+# We are using Postgres functions instead of pure stored procedures
+# These constants should escape \d sequences to ensure migration is applied
+# correctly. Manual insertion of these functions will need to unescape.
+GET_LATEST_COMPONENT_STOREDPROC_SQL = f"""
```

Comment:
This modularity seems superfluous to me eg:

1) what does modularity in python/django give us ? for me it obfuscates

2) **get_latest_components** will 'go away' the moment we get into graphdb ... we will continue to use rpmver funcs

alternately we could modularise in pg itself ...  but as the lifetime of get_latest_component is short ... not really worth the effort.

Reply:
sorry for late reply - I am fine either way - just highlighting that sacrificing comprehension for maintenance shows up as a constant 'cost' for any person reading the code ... my personal bias is towards comprehension eg. I will always choose a little bit of duplication and a little less reuse for instant comprehension most days.